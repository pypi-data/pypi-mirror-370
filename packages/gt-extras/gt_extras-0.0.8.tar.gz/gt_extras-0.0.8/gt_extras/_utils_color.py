from __future__ import annotations

import warnings

from great_tables._data_color.base import (
    _get_domain_factor,
    _html_color,
    _rescale_factor,
)
from great_tables._data_color.constants import ALL_PALETTES, DEFAULT_PALETTE
from great_tables._data_color.palettes import GradientPalette
from great_tables._tbl_data import TblData

__all__ = ["_get_discrete_colors_from_palette"]


def _get_discrete_colors_from_palette(
    palette: list[str] | str | None,
    data: list,
    data_table: TblData,
) -> list[str]:
    palette = _get_palette(palette)

    # Rescale the category column for the purpose of assigning colors to each dot
    domain_factor = _get_domain_factor(df=data_table, vals=data)
    scaled_vals = _rescale_factor(
        df=data_table,
        vals=data,
        domain=domain_factor,  # type: ignore
        palette=palette,
    )

    # Create a color scale function from the palette
    color_scale_fn = GradientPalette(colors=palette)

    # Call the color scale function on the scaled categoy values to get a list of colors
    color_vals = color_scale_fn(scaled_vals)

    for i, c in enumerate(color_vals):
        if c is None:
            color_vals[i] = "transparent"
            warnings.warn(
                "A color value is None and has been coerced to 'transparent'",
                UserWarning,
            )

    # This is redundant but satisfies the type-checker
    color_vals = [c for c in color_vals if c is not None]

    return color_vals


def _get_palette(palette: list[str] | str | None) -> list[str]:
    # If palette is not provided, use a default palette
    if palette is None:
        palette = DEFAULT_PALETTE

    # Otherwise get the palette from great_tables._data_color
    elif isinstance(palette, str):
        palette = ALL_PALETTES.get(palette, [palette])

    # Standardize values in `palette` to hexadecimal color values
    palette = _html_color(colors=palette)

    return palette
