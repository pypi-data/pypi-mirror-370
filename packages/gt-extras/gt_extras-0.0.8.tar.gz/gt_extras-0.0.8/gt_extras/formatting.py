from __future__ import annotations

import re
import tempfile
import webbrowser
from pathlib import Path
from typing import TYPE_CHECKING, Literal

import narwhals.stable.v1 as nw
from great_tables import GT
from great_tables._export import _create_temp_file_server
from great_tables._gt_data import Boxhead, ColInfo
from great_tables._tbl_data import SelectExpr, is_na

from gt_extras._utils_column import _validate_and_get_single_column

__all__ = ["fmt_pct_extra", "gt_duplicate_column", "gt_two_column_layout"]


def fmt_pct_extra(
    gt: GT,
    columns: SelectExpr,
    scale: float = 100,
    threshold: float = 1.0,
    color: str = "grey",
    decimals: int = 1,
) -> GT:
    # TODO: would we want to infer the range? (as we do in gt_plt_bar_pct)
    """
    Convert to percent and show less than `1%` as `<1%` in grey.

    The `fmt_pct_extra()` function takes an existing `GT` object and formats a column of numeric
    values as percentages. Values below the specified threshold are displayed as `"<threshold%"`
    instead of their actual percentage value, and in a color of your choosing. The default behavior
    expects values in the range `[0, 1]`, use `scale` if your values are in a different range.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    columns
        The columns containing numeric values to format as percentages.

    scale
        Multiplication factor to convert values to percentages. For some common cases:
        use `100` if values are decimals `(0.05 -> 5%)` (default),
        use `1` if values are already percentages `(5 -> 5%)`.

    threshold
        The percentage threshold below which values are displayed as `"<threshold%"` instead of
        their actual value. Note this refers to the scaled value, not the original, so you will
        have to provide `5` for `<5%` even if your original are in the range `[0,1]`.

    color
        The color to use for values below the threshold.

    decimals
        Number of decimal places to display for percentages.

    Returns
    -------
    GT
        A `GT` object with formatted percentage column.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import towny
    import gt_extras as gte

    towny_mini = towny[
        [
            "name",
            "pop_change_1996_2001_pct",
            "pop_change_2001_2006_pct",
            "pop_change_2006_2011_pct",
        ]
    ].tail(10)

    gt = (
        GT(towny_mini)
        .tab_spanner(label="Population Change", columns=[1, 2, 3])
        .cols_label(
            pop_change_1996_2001_pct="'96-'01",
            pop_change_2001_2006_pct="'01-'06",
            pop_change_2006_2011_pct="'06-'11",
        )
    )

    gt.pipe(
        gte.fmt_pct_extra,
        columns=[1, 2, 3],
        threshold=5,
        color="green",
    )
    ```
    """
    # TODO: consider how to handle negative values

    def _fmt_pct_single_val(value: float):
        if is_na(gt._tbl_data, value):
            return ""

        # Convert to percentage
        pct_value = value * scale

        if abs(pct_value) < threshold:
            return f"<span style='color:{color};'><{threshold:g}%</span>"
        else:
            return f"{pct_value:.{decimals}f}%"

    res = gt
    res = res.fmt(_fmt_pct_single_val, columns=columns)

    return res


def gt_duplicate_column(
    gt: GT,
    column: SelectExpr,
    after: str | None = None,
    append_text: str | None = "_dupe",
    dupe_name: str | None = None,
) -> GT:
    """
    Duplicate a column in a `GT` object.

    The `gt_duplicate_column()` function takes an existing `GT` object and creates a duplicate
    _(without styling)_ of the specified column. The duplicated column can be renamed using either
    `dupe_name` or by appending text to the original column name, and positioned at a
    specific location in the table.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    column
        The column to duplicate. Can be a column name or index.

    after
        The column after which to place the duplicated column. If `None`, the duplicated
        column will be moved to the end of the table.

    append_text
        Text to append to the original column name for the duplicate. Only used if
        `dupe_name` is not provided. Defaults to `"_dupe"`.

    dupe_name
        The name for the duplicated column. If provided, this overrides `append_text`.

    Returns
    -------
    GT
        A `GT` object with the duplicated column added.

    Examples
    --------
    ```{python}
    from great_tables import GT
    from great_tables.data import gtcars
    import gt_extras as gte

    gtcars_mini = gtcars[["mfr", "model", "year", "hp"]].head(5)
    gt = GT(gtcars_mini)

    # Duplicate with custom name and position
    gt.pipe(
        gte.gt_duplicate_column,
        column="hp",
        after="year",
    )
    ```

    Note
    --------
    Styles generated by [`GT.tab_style()`](https://posit-dev.github.io/great-tables/reference/GT.tab_style)
    will render in the duplicated column. This may be accommodated in the future.
    """
    original_name, _ = _validate_and_get_single_column(gt, column)

    # If dupe_name is given, it overrides append_text
    append_text = append_text or "_dupe"
    if dupe_name is not None:
        new_col_name = dupe_name
    else:
        new_col_name = original_name + append_text

    if new_col_name == original_name:
        raise ValueError(
            f"The new column name '{new_col_name}' cannot be the same as the original column name '{original_name}'."
        )

    res = gt

    # get the boxhead info
    original_col_info = None
    for col_info in res._boxhead:
        if col_info.var == original_name:
            original_col_info = col_info
            break

    if TYPE_CHECKING:
        if original_col_info is None:
            raise ValueError(f"Column '{original_name}' not found in boxhead.")

    # make the new boxhead entry
    new_col_info = ColInfo(
        var=new_col_name,
        type=original_col_info.type,
        column_label=new_col_name,
        column_align=original_col_info.column_align,
        column_width=original_col_info.column_width,
    )

    # Duplicate the col in the data_table data
    original_data_table = res._tbl_data
    new_data_table = nw.to_native(
        nw.from_native(original_data_table).with_columns(
            nw.col(original_name).alias(new_col_name)
        )
    )

    # Duplicate the col in the body element of body
    new_body = res._body
    new_body.body = nw.to_native(
        nw.from_native(new_body.body).with_columns(
            nw.col(original_name).alias(new_col_name)
        )
    )

    new_boxhead_list = list(res._boxhead._d) + [new_col_info]
    new_boxhead = Boxhead(new_boxhead_list)

    res = res._replace(_tbl_data=new_data_table, _boxhead=new_boxhead, _body=new_body)

    if after is None:
        res = res.cols_move_to_end(new_col_name)
    else:
        res = res.cols_move(new_col_name, after=after)

    return res


class GTCombinedLayout:
    """
    Represents the combined layout of two `GT` objects rendered side by side.

    This class is returned by [`gt_two_column_layout()`](https://posit-dev.github.io/gt-extras/reference/gt_two_column_layout)
    and is designed for automatic rendering in Jupyter notebooks and other HTML-aware environments.

    Parameters
    ----------
    html_content
        The combined HTML content representing the two tables in a single layout.

    Methods
    -------
    _repr_html_()
        Returns the HTML content for notebook rendering.

    __str__()
        Returns the HTML content as a string.

    show()
        Displays the table in an environment of choice. Options are `"notebook"` and `"browser"`.
    """

    def __init__(self, html_content: str):
        self.html_content = html_content

    def _repr_html_(self):
        return self.html_content

    def __str__(self):
        return self.html_content

    # do a show based on code from great_tables `GT.show()`
    def show(self, target: Literal["notebook", "browser"] = "browser"):
        if target == "notebook":
            from IPython.core.display import display_html

            # https://github.com/ipython/ipython/pull/10962
            display_html(  # pyright: ignore[reportUnknownVariableType]
                self._repr_html_(), raw=True, metadata={"text/html": {"isolated": True}}
            )
        elif target == "browser":
            with tempfile.TemporaryDirectory() as tmp_dir:
                f_path = Path(tmp_dir) / "index.html"
                f_path.write_text(self._repr_html_(), encoding="utf-8")

                # create a server that closes after 1 request ----
                server = _create_temp_file_server(f_path)
                webbrowser.open(f"http://127.0.0.1:{server.server_port}/{f_path.name}")
                server.handle_request()
        else:
            raise Exception(f"Unknown target display: {target}")

    def save(self):
        raise NotImplementedError


def gt_two_column_layout(
    gt1: GT,
    gt2: GT,
    target: Literal["notebook", "browser"] | None = None,  # TODO: add save
    table_header_from: Literal[1, 2] | None = None,
) -> GTCombinedLayout:
    """
    Combine two `GT` objects into a two-column layout.

    This function takes two `GT` objects and arranges them side by side in a single HTML output.
    Optionally, the header (title and subtitle) from one of the tables can be included at the top
    of the combined layout. The output can be displayed directly in a Jupyter notebook, opened in
    a browser, or returned as HTML.

    Parameters
    ----------
    gt1
        The first table to include in the layout. This table will appear on the left side.

    gt2
        The second table to include in the layout. This table will appear on the right side.

    target
        Determines how the combined layout is displayed. Use `"notebook"` to display in a Jupyter
        notebook, `"browser"` to open in a web browser. If `None`, returns the layout object
        for automatic rendering. Currently, `"save"` is not yet implemented.

    table_header_from
        Determines which table's header (title and subtitle) to include in the combined layout.
        If set to `1`, the header from `gt1` will be used. If set to `2`, the header from `gt2`
        will be used. If `None`, no header will be included.

    Returns
    -------
    GTCombinedLayout
        A layout object that automatically renders as HTML in Jupyter notebooks and Quarto
        documents. It also has a `.show()` method, taking `"notebook"` or `"browser"` as targets.

    Examples
    --------
    ```{python}
    from great_tables import GT
    import gt_extras as gte
    import pandas as pd

    df1 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
    df2 = pd.DataFrame({"X": [7, 8, 9], "Y": [10, 11, 12]})

    gt1 = GT(df1).tab_header(title="Table 1", subtitle="1st Table")
    gt2 = GT(df2).tab_header(title="Table 2", subtitle="2nd Table")

    gte.gt_two_column_layout(gt1, gt2)
    ```

    Alternatively, we can combine data from the same table in a two column layout like so:

    ```{python}
    df1 = pd.DataFrame({"A": [1, 2, 3], "B": [4, 5, 6]})
    df2 = pd.DataFrame({"A": [7, 8, 9], "B": [10, 11, 12]})

    gt1 = GT(df1).tab_header(title="Joined Table")
    gt2 = GT(df2)

    gte.gt_two_column_layout(gt1, gt2, table_header_from=1)
    ```


    """

    def extract_tab_header_and_style(gt: GT) -> dict:
        """
        Extract the title, subtitle, and full style block from a GT object's HTML.
        """
        html = gt.as_raw_html()

        title = gt._heading.title if gt._heading.title else ""
        subtitle = gt._heading.subtitle if gt._heading.subtitle else ""

        # Extract the original table ID
        id_match = re.search(r'<div id="([^"]*)"', html)
        original_id = id_match.group(1) if id_match else ""

        # Extract the entire style block
        style_match = re.search(r"<style>(.*?)</style>", html, re.DOTALL)
        style_content = style_match.group(1) if style_match else ""

        # Replace the original ID with "mycombinedtable" in the CSS
        if original_id:
            style_content = style_content.replace(f"#{original_id}", "#mycombinedtable")
            style_content = style_content.replace(
                "mycombinedtable table", "mycombinedtable div"
            )

        title_td_pattern = (
            r'<td[^>]*class="[^"]*gt_title[^"]*"[^>]*>\s*'
            + re.escape(str(title))
            + r"\s*</td>"
        )
        title_match = re.search(title_td_pattern, html, re.DOTALL)

        if title_match:
            td_tag = title_match.group(0)

            # Extract class attribute
            class_match = re.search(r'class="([^"]*gt_title[^"]*)"', td_tag)
            title_class = f"gt_table {class_match.group(1)}" if class_match else ""

            # Extract style attribute
            style_match = re.search(r'style="([^"]*)"', td_tag)
            title_inline_style = style_match.group(1) if style_match else ""
        else:
            title_class = ""
            title_inline_style = ""

        subtitle_class = ""

        subtitle_td_pattern = (
            r'<td[^>]*class="[^"]*gt_subtitle[^"]*"[^>]*>\s*'
            + re.escape(str(subtitle))
            + r"\s*</td>"
        )
        subtitle_match = re.search(subtitle_td_pattern, html, re.DOTALL)

        if subtitle_match:
            td_tag = subtitle_match.group(0)

            # Extract class attribute
            class_match = re.search(r'class="([^"]*gt_subtitle[^"]*)"', td_tag)

            if class_match:
                # Exclude gt_bottom_border from class string if present
                class_str = class_match.group(1)
                class_str = " ".join(
                    c for c in class_str.split() if c != "gt_bottom_border"
                )
                subtitle_class = f"gt_table {class_str}"

            # Extract style attribute
            style_match = re.search(r'style="([^"]*)"', td_tag)
            subtitle_inline_style = style_match.group(1) if style_match else ""
        else:
            subtitle_inline_style = ""

        return {
            "title": title,
            "subtitle": subtitle,
            "title_class": title_class,
            "subtitle_class": subtitle_class,
            "title_style": title_inline_style,
            "subtitle_style": subtitle_inline_style,
            "style": style_content,
        }

    # Extract header data if specified
    if table_header_from == 1:
        header_data = extract_tab_header_and_style(gt1)
    elif table_header_from == 2:
        header_data = extract_tab_header_and_style(gt2)
    else:
        header_data = None

    # Build the header HTML if header data is available
    if header_data and (header_data["title"] or header_data["subtitle"]):
        header_html = f"""
        <style>
        {header_data["style"]}
        </style>
        """

        # Add title if it exists
        if header_data["title"]:
            header_html += f"""
            <div class="{header_data["title_class"]}"
                style="width:100%; {header_data["title_style"]}"
            >
                {header_data["title"]}
            </div>
            """

        # Add subtitle if it exists
        if header_data["subtitle"]:
            header_html += f"""
            <div class="{header_data["subtitle_class"]}"
                style="width:100%; {header_data["subtitle_style"]}">
                {header_data["subtitle"]}
            </div>
            """

        # Remove headers from the two tables
        gt1_no_header = gt1.tab_header(title=None, subtitle=None)  # type: ignore
        table_1_html = gt1_no_header.as_raw_html()
        gt2_no_header = gt2.tab_header(title=None, subtitle=None)  # type: ignore
        table_2_html = gt2_no_header.as_raw_html()

    else:
        header_html = ""
        table_1_html = gt1.as_raw_html()
        table_2_html = gt2.as_raw_html()

    double_table_html = f"""
    <div style="display: flex; justify-content: center; width: 100%;">
        <div id="mycombinedtable" style="display: inline-block; width: auto;">
            {header_html}
            <div style="overflow: auto; white-space: nowrap;">
                <div style="display: inline-block; margin-right: 1em;">
                    {table_1_html}
                </div>
                <div style="display: inline-block;">
                    {table_2_html}
                </div>
            </div>
        </div>
    </div>
    """

    double_table = GTCombinedLayout(double_table_html)

    if target:
        if target == "save":
            raise NotImplementedError(
                "At the moment, only notebook and browser display options are available."
            )
        else:
            double_table.show(target)

    return GTCombinedLayout(double_table_html)
