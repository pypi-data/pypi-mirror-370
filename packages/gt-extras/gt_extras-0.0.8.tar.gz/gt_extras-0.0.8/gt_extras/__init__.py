# Import objects from the module
from .colors import (
    gt_color_box,
    gt_data_color_by_group,
    gt_highlight_cols,
    gt_highlight_rows,
    gt_hulk_col_numeric,
)
from .formatting import fmt_pct_extra, gt_duplicate_column, gt_two_column_layout
from .html import gt_merge_stack, with_hyperlink, with_tooltip
from .icons import fa_icon_repeat, gt_fa_rank_change, gt_fa_rating
from .images import add_text_img, gt_fmt_img_circle, img_header
from .plotting import (
    gt_plt_bar,
    gt_plt_bar_pct,
    gt_plt_bar_stack,
    gt_plt_bullet,
    gt_plt_conf_int,
    gt_plt_donut,
    gt_plt_dot,
    gt_plt_dumbbell,
    gt_plt_winloss,
)
from .styling import gt_add_divider
from .summary import gt_plt_summary
from .themes import (
    gt_theme_538,
    gt_theme_dark,
    gt_theme_dot_matrix,
    gt_theme_espn,
    gt_theme_excel,
    gt_theme_guardian,
    gt_theme_nytimes,
    gt_theme_pff,
)

__all__ = [
    "gt_theme_538",
    "gt_theme_espn",
    "gt_theme_nytimes",
    "gt_theme_guardian",
    "gt_theme_excel",
    "gt_theme_dot_matrix",
    "gt_theme_dark",
    "gt_theme_pff",
    "gt_data_color_by_group",
    "gt_highlight_cols",
    "gt_highlight_rows",
    "gt_hulk_col_numeric",
    "gt_color_box",
    "fa_icon_repeat",
    "gt_fa_rank_change",
    "gt_fa_rating",
    "gt_plt_bar",
    "gt_plt_bar_pct",
    "gt_plt_bullet",
    "gt_plt_dot",
    "gt_plt_conf_int",
    "gt_plt_dumbbell",
    "gt_plt_donut",
    "gt_plt_winloss",
    "gt_plt_bar_stack",
    "with_hyperlink",
    "with_tooltip",
    "gt_merge_stack",
    "fmt_pct_extra",
    "gt_duplicate_column",
    "gt_two_column_layout",
    "add_text_img",
    "img_header",
    "gt_fmt_img_circle",
    "gt_add_divider",
    "gt_plt_summary",
]
