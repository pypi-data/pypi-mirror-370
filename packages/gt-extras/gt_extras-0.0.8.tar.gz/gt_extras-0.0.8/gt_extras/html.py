from __future__ import annotations

from typing import Literal

from great_tables import GT
from great_tables._tbl_data import SelectExpr, is_na

from gt_extras._utils_column import _validate_and_get_single_column

__all__ = ["with_hyperlink", "with_tooltip", "gt_merge_stack"]


def with_hyperlink(text: str, url: str, new_tab: bool = True) -> str:
    """
    Create HTML hyperlinks for use in `GT` cells.

    The `with_hyperlink()` function creates properly formatted HTML hyperlink elements that can be
    used within table cells.

    Parameters
    ----------
    text
        A string that will be displayed as the clickable link text.

    url
        A string indicating the destination URL for the hyperlink.

    new_tab
        A boolean indicating whether the link should open in a new browser tab or the current tab.

    Returns
    -------
    str
        An string containing the HTML formatted hyperlink element.

    Examples
    --------
    ```{python}
    import pandas as pd
    from great_tables import GT
    import gt_extras as gte

    df = pd.DataFrame(
        {
            "name": ["Great Tables", "Plotnine", "Quarto"],
            "url": [
                "https://posit-dev.github.io/great-tables/",
                "https://plotnine.org/",
                "https://quarto.org/",
            ],
            "github_stars": [2334, 4256, 4628],
            "repo_url": [
                "https://github.com/posit-dev/great-tables",
                "https://github.com/has2k1/plotnine",
                "https://github.com/quarto-dev/quarto-cli",
            ],
        }
    )

    df["Package"] = [
        gte.with_hyperlink(name, url)
        for name, url in zip(df["name"], df["url"])
    ]

    df["Github Stars"] = [
        gte.with_hyperlink(github_stars, repo_url, new_tab=False)
        for github_stars, repo_url in zip(df["github_stars"], df["repo_url"])
    ]

    GT(df[["Package", "Github Stars"]])
    ```
    """
    target = "_self"
    if new_tab:
        target = "_blank"

    return f'<a href="{url}" target="{target}">{text}</a>'


def with_tooltip(
    label: str,
    tooltip: str,
    text_decoration_style: Literal["solid", "dotted", "none"] = "dotted",
    color: str | Literal["none"] = "blue",
) -> str:
    """
    Create HTML text with tooltip functionality for use in `GT` cells.

    The `with_tooltip()` function creates an HTML `<abbr>` element with a tooltip that appears
    when users hover over the text. The text can be styled with customizable underline styles
    and colors to indicate it's interactive.

    Parameters
    ----------
    label
        A string that will be displayed as the visible text.

    tooltip
        A string that will appear as the tooltip when hovering over the label.

    text_decoration_style
        A string indicating the style of underline decoration. Options are `"solid"`,
        `"dotted"`, or "none".

    color
        A string indicating the text color. If "none", no color styling is applied.

    Returns
    -------
    str
        An HTML string containing the formatted tooltip element.

    Examples
    -------
    ```{python}
    import pandas as pd
    from great_tables import GT
    import gt_extras as gte

    df = pd.DataFrame(
        {
            "name": ["Great Tables", "Plotnine", "Quarto"],
            "description": [
                "Absolutely Delightful Table-making in Python",
                "A grammar of graphics for Python",
                "An open-source scientific and technical publishing system",
            ],
        }
    )

    df["Package"] = [
        gte.with_tooltip(name, description, color = "none")
        for name, description in zip(df["name"], df["description"])
    ]

    GT(df[["Package"]])
    ```
    """

    # Throw if `text_decoration_style` is not one of the allowed values
    if text_decoration_style not in ["none", "solid", "dotted"]:
        raise ValueError(
            "Text_decoration_style must be one of 'none', 'solid', or 'dotted'"
        )

    if color is None:
        raise ValueError("color must be a string or 'none', not None.")

    style = "cursor: help; "

    if text_decoration_style != "none":
        style += "text-decoration: underline; "
        style += f"text-decoration-style: {text_decoration_style}; "
    else:
        style += "text-decoration: none; "

    if color != "none":
        style += f"color: {color}; "

    # Why doesn't the output have to be wrapped in GT.html()?
    return f'<abbr style="{style}" title="{tooltip}">{label}</abbr>'


def gt_merge_stack(
    gt: GT,
    col1: SelectExpr,
    col2: SelectExpr,
    font_size_main: int = 14,
    font_size_secondary: int = 10,
    font_weight_main: Literal["normal", "bold", "bolder", "lighter"] | int = "bold",
    font_weight_secondary: Literal["normal", "bold", "bolder", "lighter"]
    | int = "normal",
    color_main: str = "black",
    color_secondary: str = "grey",
    small_caps: bool = True,
) -> GT:
    """
    Merge two columns into a stacked format within a `GT` object.

    The `gt_merge_stack()` function combines two columns in a `GT` object into a single column
    with a stacked format. The top section displays values from the first column (`col1`), and
    the bottom section displays values from the second column (`col2`). Both sections can be
    styled independently with customizable font sizes, weights, colors, and text variants.

    The resulting table will hide `col2`, and the orignal `col1` will contain the merged entries.

    Parameters
    ----------
    gt
        A `GT` object to modify.

    col1
        The column containing values to display in the top section of the stack.

    col2
        The column containing values to display in the bottom section of the stack.

    font_size_main
        The font size for the top section of the stack.

    font_size_secondary
        The font size for the bottom section of the stack.

    font_weight_main
        The font weight for the top section of the stack. Options include `"normal"`, `"bold"`,
        `"bolder"`, `"lighter"`, or an integer value.

    font_weight_secondary
        The font weight for the bottom section of the stack.

    color_main
        The text color for the top section of the stack.

    color_secondary
        The text color for the bottom section of the stack.

    small_caps
        A boolean indicating whether the top section should use small caps styling.

    Returns
    -------
    GT
        A `GT` object with the merged and styled column.

    Examples
    -------
    ```{python}
    import pandas as pd
    from great_tables import GT
    import gt_extras as gte

    df = pd.read_csv("../assets/teams_colors_logos.csv")
    df = (df.filter(items=["team_nick", "team_abbr", "team_conf", "team_division", "team_wordmark"]).head(8))

    gt = GT(df, groupname_col="team_conf", rowname_col="team_nick")
    gt = gt.fmt_image(columns="team_wordmark")


    gt.pipe(
        gte.gt_merge_stack,
        col1="team_nick",
        col2="team_division",
    )
    ```
    """

    def _make_merge_stack_html(
        col1_val: str,
        col2_val: str,
        font_size_main: int,
        font_size_secondary: int,
        font_weight_main: str | int,
        font_weight_secondary: str | int,
        color_main: str,
        color_secondary: str,
        small_caps: bool,
    ) -> str:
        font_variant = "small-caps" if small_caps else "normal"

        top_section_html = f"""
        <div style="line-height:{font_size_main}px;">
            <span style="
                font-weight:{font_weight_main};
                font-variant:{font_variant};
                color:{color_main};
                font-size:{font_size_main}px;
            ">
                {col1_val}
            </span>
        </div>
        """.strip()

        bottom_section_html = f"""
        <div style="line-height:{font_size_secondary}px;">
            <span style="
                font-weight:{font_weight_secondary};
                color:{color_secondary};
                font-size:{font_size_secondary}px;
            ">
                {col2_val}
            </span>
        </div>
        """.strip()

        html = f"""
        <div>
            {top_section_html}
            {bottom_section_html}
        </div>
        """.strip()

        return html

    _, col1_vals = _validate_and_get_single_column(gt, expr=col1)
    _, col2_vals = _validate_and_get_single_column(gt, expr=col2)

    res = gt

    for i in range(len(gt._tbl_data)):
        col1_val = col1_vals[i]
        col2_val = col2_vals[i]

        if is_na(gt._tbl_data, col1_val):
            col1_val = ""
        if is_na(gt._tbl_data, col2_val):
            col2_val = ""

        res = res.fmt(
            lambda _, col1_val=col1_val, col2_val=col2_val: _make_merge_stack_html(
                col1_val=col1_val,
                col2_val=col2_val,
                font_size_main=font_size_main,
                font_size_secondary=font_size_secondary,
                font_weight_main=font_weight_main,
                font_weight_secondary=font_weight_secondary,
                color_main=color_main,
                color_secondary=color_secondary,
                small_caps=small_caps,
            ),
            columns=col1,
            rows=[i],
        )

    res = res.cols_hide(col2)

    return res
