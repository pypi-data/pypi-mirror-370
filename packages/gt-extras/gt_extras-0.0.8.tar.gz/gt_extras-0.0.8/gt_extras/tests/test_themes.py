from gt_extras import (
    gt_theme_538,
    gt_theme_espn,
    gt_theme_nytimes,
    gt_theme_guardian,
    gt_theme_excel,
    gt_theme_dot_matrix,
    gt_theme_dark,
    gt_theme_pff,
)

from conftest import (
    assert_rendered_global_imports,
)


def test_theme_538_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_538(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_espn_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_espn(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_nytimes_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_nytimes(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_guardian_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_guardian(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_excel_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_excel(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_excel_color(mini_gt):
    themed_gt = gt_theme_excel(gt=mini_gt, color="lightblue")
    html = themed_gt.as_raw_html()
    assert "lightblue" in html


def test_theme_dot_matrix_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_dot_matrix(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_dot_matrix_color(mini_gt):
    themed_gt = gt_theme_dot_matrix(gt=mini_gt, color="lightblue")
    html = themed_gt.as_raw_html()
    assert "lightblue" in html


def test_theme_dark_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_dark(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_pff_fonts_snap(snapshot, mini_gt):
    themed_gt = gt_theme_pff(gt=mini_gt)
    assert_rendered_global_imports(snapshot, themed_gt)


def test_theme_pff_dividers(mini_gt):
    themed_gt = gt_theme_pff(gt=mini_gt, divider="num")
    html = themed_gt.as_raw_html()
    assert "border-left: 2px solid lightgrey" in html


def test_theme_pff_spanner(mini_gt):
    gt_with_spanner = mini_gt.tab_spanner("Spanner Label", columns=["num", "char"])
    themed_gt = gt_theme_pff(gt=gt_with_spanner, spanners=["Spanner Label"])
    html = themed_gt.as_raw_html()
    assert "Spanner Label" in html
    # This assertion ensures the blank spanner is created
    assert '<span class="gt_column_spanner"> </span>' in html


def test_theme_pff_rank_col(mini_gt):
    themed_gt = gt_theme_pff(gt=mini_gt, rank_col="num")
    html = themed_gt.as_raw_html()
    assert "#e4e8ec" in html
