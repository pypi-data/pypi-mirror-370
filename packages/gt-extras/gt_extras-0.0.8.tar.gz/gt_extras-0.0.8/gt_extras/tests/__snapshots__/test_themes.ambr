# serializer version: 1
# name: test_theme_538_fonts_snap
  '''
  @import url('https://fonts.googleapis.com/css2?family=Cairo&display=swap');
  @import url('https://fonts.googleapis.com/css2?family=Chivo&display=swap');
  '''
# ---
# name: test_theme_dark_fonts_snap
  '''
  @import url('https://fonts.googleapis.com/css2?family=Libre+Franklin&display=swap');
  @import url('https://fonts.googleapis.com/css2?family=Source+Sans+Pro&display=swap');
  '''
# ---
# name: test_theme_dot_matrix_fonts_snap
  ''
# ---
# name: test_theme_espn_fonts_snap
  "@import url('https://fonts.googleapis.com/css2?family=Lato&display=swap');"
# ---
# name: test_theme_excel_fonts_snap
  ''
# ---
# name: test_theme_guardian_fonts_snap
  "@import url('https://fonts.googleapis.com/css2?family=Noto+Sans&display=swap');"
# ---
# name: test_theme_nytimes_fonts_snap
  '''
  @import url('https://fonts.googleapis.com/css2?family=Libre+Franklin&display=swap');
  @import url('https://fonts.googleapis.com/css2?family=Source+Sans+Pro&display=swap');
  '''
# ---
# name: test_theme_pff_fonts_snap
  "@import url('https://fonts.googleapis.com/css2?family=Roboto&display=swap');"
# ---
